src/pkg/packager/deploy.go

```diff
@@ -82,13 +83,45 @@ func (p *Packager) Deploy() error {
 
 	// Filter out components that are not compatible with this system
 	p.filterComponents(true)
+	componentsToDeploy := p.getValidComponents()
+
+	// Check if any of the components we are going to deploy require a cluster
+	requiresCluster := false
+	for _, component := range componentsToDeploy {
+		requiresCluster = requiresCluster || component.RequiresCluster()
+	}
+
+	// If we require the cluster for a regular package deployment, check if a hook has been uploaded to the cluster
+	// NOTE: We are not doing this for init packages at this time, init packages will manage hooks on their own.
+	if requiresCluster && !p.cfg.IsInitConfig {
+		// Make sure we have access to the cluster
+		var err error
+		if p.cluster == nil {
+			p.cluster, err = cluster.NewClusterWithWait(cluster.DefaultTimeout, true)
+			if err != nil {
+				return fmt.Errorf("unable to connect to the Kubernetes cluster: %w", err)
+			}
+		}
 
-	// Get a list of all the components we are deploying and actually deploy them
-	deployedComponents, err := p.deployComponents()
+		if err := p.getAllClusterHooks(); err != nil {
+			return fmt.Errorf("unable to get all cluster hooks: %w", err)
+		}
+
+		if err := p.runPreDeployHooks(); err != nil {
+			return fmt.Errorf("unable to run pre-deploy hooks: %w", err)
+		}
+	}
+
+	// Deploy all of the components
+	deployedComponents, err := p.deployComponents(componentsToDeploy)
```

Comment:
We may want to drop "package" hooks and go with "runonce" component hooks instead

Reply:
I am not going to do this right now, I am leaving a TODO comment for this and I"ll be sure to discuss the different approaches in the ADR that we are working on.